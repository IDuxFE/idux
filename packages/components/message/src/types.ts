import type { DefineComponent, HTMLAttributes, VNode } from 'vue'
import type { IxInnerPropTypes, IxPublicPropTypes } from '@idux/cdk/utils'

import { IxPropTypes } from '@idux/cdk/utils'

export type MessageType = 'info' | 'success' | 'warning' | 'error' | 'loading'

export interface MessageOptions extends MessagePublicProps {
  id?: string
  content?: string | VNode
  onDestroy?: () => void
}
export interface MessageRef {
  id: string
  update: (options: MessageOptions) => void
  destroy: () => void
}

export const messageProps = {
  visible: IxPropTypes.bool.def(false),
  destroyOnHover: IxPropTypes.bool,
  duration: IxPropTypes.number,
  icon: IxPropTypes.oneOfType([String, IxPropTypes.vNode]),
  type: IxPropTypes.oneOf<MessageType>(['info', 'success', 'warning', 'error', 'loading']).def('info'),

  // events
  'onUpdate:visible': IxPropTypes.emit<(visible: boolean) => void>(),
}

export type MessageProps = IxInnerPropTypes<typeof messageProps>
export type MessagePublicProps = IxPublicPropTypes<typeof messageProps>
export type MessageComponent = DefineComponent<HTMLAttributes & typeof messageProps>
export type MessageInstance = InstanceType<DefineComponent<MessageProps>>

export const messageProviderProps = {
  maxCount: IxPropTypes.number,
  top: IxPropTypes.oneOfType([String, Number]),
}
export interface MessageProviderRef {
  open: (options: MessageOptions) => MessageRef
  info: (content: string | VNode, options?: Omit<MessageOptions, 'type' | 'content'>) => MessageRef
  success: (content: string | VNode, options?: Omit<MessageOptions, 'type' | 'content'>) => MessageRef
  warning: (content: string | VNode, options?: Omit<MessageOptions, 'type' | 'content'>) => MessageRef
  error: (content: string | VNode, options?: Omit<MessageOptions, 'type' | 'content'>) => MessageRef
  loading: (content: string | VNode, options?: Omit<MessageOptions, 'type' | 'content'>) => MessageRef
  update: (id: string, options: MessageOptions) => void
  destroy: (id: string | string[]) => void
  destroyAll: () => void
}

export type MessageProviderProps = IxInnerPropTypes<typeof messageProviderProps>
export type MessageProviderPublicProps = IxPublicPropTypes<typeof messageProviderProps>
export type MessageProviderComponent = DefineComponent<typeof messageProviderProps, MessageProviderRef>
export type MessageProviderInstance = InstanceType<DefineComponent<MessageProviderProps, MessageProviderRef>>
