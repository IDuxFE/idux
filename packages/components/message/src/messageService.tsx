import { useContainer } from '@idux/cdk/portal'
import { isString, toCssPixel, uniqueId, isFunction, isUndefined } from '@idux/cdk/utils'
import { createVNode, render, ref, isVNode, VNode, TransitionGroup } from 'vue'
import { IxPortal } from '@idux/cdk/portal'
import IxMessage from './Message'
import { MessageId, MessageListItem, MessageApi, MessageApiFnReturn, MessageType } from './types'

const containerCls = 'ix-message'
let isInit = false
const maxCount = 5 // todo 全局设置获取
const top = 60 // todo 全局设置获取
const { messages, add, remove } = useMessage()

const create = (option: Partial<MessageListItem>): MessageApiFnReturn => {
  if (!isInit) {
    const container = useContainer(containerCls)
    const onClose = (id: MessageId) => {
      remove(id)
    }
    const vm = createVNode(
      IxPortal,
      {
        target: containerCls,
      },
      {
        default: () => {
          const messageList = messages.value.map((message, index) => (
            <IxMessage
              key={message.messageId}
              duration={message.duration}
              pauseOnHover={message.pauseOnHover}
              messageId={message.messageId as MessageId}
              icon={message.icon}
              content={message.content}
              type={message.type}
              onClose={onClose}>
              {index}
            </IxMessage>
          ))
          return <TransitionGroup tag={'div'} name={'move-up'}>
            {messageList}
          </TransitionGroup>
        },
      },
    )
    render(vm, container)
    container.style.top = toCssPixel(top)
  }
  isInit = true
  const newMessageId = add(option)
  return {
    messageId: newMessageId,
    destroy: () => remove(newMessageId),
  }
}
const api: any = {
  remove,
  create,
  getContainer(): HTMLElement {
    return useContainer(containerCls)
  },
}

function useMessage() {
  const messages = ref<Partial<MessageListItem>[]>([])
  const getCurIndex = (id: MessageId | undefined) => {
    return messages.value.findIndex(message => message.messageId === id)
  }
  const add = (item: Partial<MessageListItem>): MessageId => {
    const curIndex = getCurIndex(item.messageId)
    const newMessageId = uniqueId('ix-message-item')
    if (curIndex !== -1) {
      messages.value.splice(curIndex, 1, item)
      return item.messageId as MessageId
    }
    if (messages.value.length >= maxCount) {
      messages.value = messages.value.slice(-maxCount + 1)
    }
    messages.value.push(
      Object.assign(
        {
          messageId: newMessageId,
        },
        item,
      ),
    )
    return newMessageId
  }
  const removeItem = (id: MessageId) => {
    if (messages.value.length) {
      const curItemIndex = getCurIndex(id)
      if (curItemIndex !== -1) {
        const curMessage = messages.value[curItemIndex]
        const messageClose = curMessage.onClose
        if (isFunction(messageClose)) {
          messageClose(id)
        }
        messages.value.splice(curItemIndex, 1)
      }
    }
  }
  const removeAll = () => {
    messages.value = []
    return Promise.resolve()
  }
  const remove = (id?: MessageId | MessageId[]) => {
    if (isUndefined(id)) {
      return removeAll()
    }
    if (Array.isArray(id)) {
      id.forEach(item => removeItem(item))
      return Promise.resolve()
    }
    removeItem(id)
    return Promise.resolve()
  }
  return {
    messages,
    add,
    remove,
  }
}

function formatOption(option: string | VNode | Partial<Omit<MessageListItem, 'type'>>, duration?: number) {
  // if option is plain string or vnode, return an object with content
  if (isString(option) || isVNode(option)) {
    return { content: option, duration }
  }
  if (isUndefined(duration)) {
    duration = option.duration
  }
  return {
    ...option,
    duration,
  }
}

['success', 'error', 'info', 'warn'].forEach((type) => {
  api[type] = function (
    option: string | Partial<Omit<MessageListItem, 'type'>> = '',
    duration?: number,
  ): MessageApiFnReturn {
    option = formatOption(option, duration)
    return create({
      ...option,
      type: type as MessageType,
    })
  }
})

export const MessageService: MessageApi = api
