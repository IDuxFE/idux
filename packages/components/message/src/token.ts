import type { InjectionKey } from 'vue'
import type { MessageProviderRef } from './types'

export const messageProviderToken: InjectionKey<MessageProviderRef> = Symbol('messageProviderToken')
