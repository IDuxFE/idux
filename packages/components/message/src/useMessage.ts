import type { MessageProviderRef } from './types'

import { inject } from 'vue'
import { throwError } from '@idux/cdk/utils'
import { messageProviderToken } from './token'

export const useMessage = (): MessageProviderRef => {
  const modalProviderRef = inject(messageProviderToken, null)
  if (modalProviderRef === null) {
    return throwError('components/message', '<ix-message-provider> not found.')
  }
  return modalProviderRef
}
