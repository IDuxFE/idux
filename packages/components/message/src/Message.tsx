import { defineComponent, onUnmounted, onMounted, isVNode, h, computed } from 'vue'
import { IxIcon } from '@idux/components/icon'
import { PropTypes, isUndefined } from '@idux/cdk/utils'
import { useGlobalConfig } from '@idux/components'
import { MessageItemProps, MessageType, MessageId } from './types'

export default defineComponent({
  name: 'IxMessageItem',
  components: {
    IxIcon,
  },
  props: {
    duration: PropTypes.number,
    pauseOnHover: PropTypes.bool,
    icon: PropTypes.oneOfType([PropTypes.string, PropTypes.object]).def(''),
    content: PropTypes.oneOfType([PropTypes.string, PropTypes.object]).def(''),
    type: PropTypes.oneOf(['success', 'error', 'info', 'warn'] as const),
    messageId: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
    onClose: PropTypes.func
  },
  setup(props: MessageItemProps) {
    // init
    const classes = useClasses(props)
    const messageConfig = useGlobalConfig('message')
    const { startCloseTimer, clearCloseTimer, destroy } = usePauseOnHover({
      duration: props.duration ?? messageConfig.duration,
      pauseOnHover: props.pauseOnHover ?? messageConfig.pauseOnHover,
      onClose: props.onClose,
      messageId: props.messageId
    })
    onMounted(() => {
      startCloseTimer()
    })
    onUnmounted(() => {
      destroy()
    })

    return {
      classes,
      startCloseTimer,
      clearCloseTimer,
    }
  },
  render() {
    const content = isVNode(this.content) ? h(this.content) : this.content
    const icon = isVNode(this.icon) ? h(this.icon) : getIcon(this.icon as string, this.type)

    return (
      <div class={this.classes} onMouseenter={this.clearCloseTimer} onMouseleave={this.startCloseTimer}>
        <div class="ix-message-item-content">
          <div class="ix-message-item-custom-content">
            {icon}
            <span class="ix-message-item-custom-text">{content}</span>
          </div>
        </div>
      </div>
    )
  },
})
function useClasses(props: MessageItemProps) {
  return computed(() => {
    const type = props.type
    const baseCls= 'ix-message-item'

    return [baseCls, `${baseCls}-${type}`]
  })
}
function getIcon(icon: string, type: MessageType | string | undefined) {
  if (isUndefined(type) || type === '') {
    return ''
  }
  const ICON_MAP = {
    success: 'check-circle',
    error: 'close-circle',
    info: 'info-circle',
    warn: 'exclamation-circle',
  }
  const mapIconName: string = icon || ICON_MAP[type as MessageType]

  return <ix-icon name={mapIconName} />
}
function usePauseOnHover(options: { duration: number; pauseOnHover: boolean, onClose: (id: MessageId) => void, messageId: MessageId }) {
  let timer: number | undefined | null
  const { duration, pauseOnHover, messageId, onClose } = options
  const startCloseTimer = () => {
    timer = window.setTimeout(() => {
      close()
    }, duration)
  }
  const clearCloseTimer = () => {
    if (pauseOnHover && timer) {
      clearTimeout(timer)
      timer = null
    }
  }
  const close = () => {
    clearCloseTimer()
    onClose(messageId)
  }
  const destroy = () => {
    if (timer) {
      clearTimeout(timer)
      timer = null
      onClose(messageId)
    }
  }
  return {
    startCloseTimer,
    clearCloseTimer,
    destroy,
  }
}
