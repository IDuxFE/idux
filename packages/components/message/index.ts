/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { MessageComponent, MessageProviderComponent } from './src/types'

import Message from './src/Message'
import MessageProvider from './src/MessageProvider'

const IxMessage = Message as unknown as MessageComponent
const IxMessageProvider = MessageProvider as unknown as MessageProviderComponent

export { IxMessage, IxMessageProvider }
export { useMessage } from './src/useMessage'

export type {
  MessageInstance,
  MessageComponent,
  MessagePublicProps as MessageProps,
  MessageProviderInstance,
  MessageProviderComponent,
  MessageProviderPublicProps as MessageProviderProps,
  MessageProviderRef,
  MessageType,
  MessageOptions,
  MessageRef,
} from './src/types'
