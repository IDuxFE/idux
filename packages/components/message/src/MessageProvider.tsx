/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import {
  type ComputedRef,
  TransitionGroup,
  type VNode,
  type VNodeChild,
  computed,
  defineComponent,
  provide,
  ref,
} from 'vue'

import { CdkPortal } from '@idux/cdk/portal'
import { VKey, callEmit, convertArray, convertCssPixel, uniqueId } from '@idux/cdk/utils'
import { useGlobalConfig } from '@idux/components/config'
import { useThemeToken } from '@idux/components/theme'
import { convertStringVNode, usePortalTarget } from '@idux/components/utils'

import Message from './Message'
import { MESSAGE_PROVIDER_TOKEN } from './token'
import { type MessageOptions, type MessageRef, messageProviderProps } from './types'
import { getThemeTokens, transforms } from '../theme'

export default defineComponent({
  name: 'IxMessageProvider',
  inheritAttrs: false,
  props: messageProviderProps,
  setup(props, { expose, slots, attrs }) {
    const common = useGlobalConfig('common')
    const { globalHashId, hashId, registerToken } = useThemeToken('message')
    registerToken(getThemeTokens, transforms)

    const config = useGlobalConfig('message')
    const mergedPrefixCls = computed(() => `${common.prefixCls}-message`)
    const mergedPortalTarget = usePortalTarget(props, config, common, mergedPrefixCls)

    const style = computed(() => ({ top: convertCssPixel(props.top ?? config.top) }))
    const maxCount = computed(() => props.maxCount ?? config.maxCount)
    const { messages, loadContainer, open, info, success, warning, error, loading, update, destroy, destroyAll } =
      useMessage(maxCount)

    const apis = { open, info, success, warning, error, loading, update, destroy, destroyAll }

    provide(MESSAGE_PROVIDER_TOKEN, apis)
    expose(apis)

    return () => {
      const child = messages.value.map(item => {
        // The default value for `visible` is true, onDestroy should not be passed in Message
        const { key, content, visible = true, onDestroy, ...restProps } = item
        const onClose = () => destroy(key!)
        const mergedProps = { key, visible, onClose }
        const contentNode = convertStringVNode(undefined, content)
        return (
          <Message {...mergedProps} {...restProps}>
            {contentNode}
          </Message>
        )
      })

      return (
        <>
          {slots.default?.()}
          <CdkPortal target={mergedPortalTarget.value} load={loadContainer.value}>
            <TransitionGroup
              tag="div"
              appear
              name={`${common.prefixCls}-move-up`}
              class={[`${mergedPrefixCls.value}-wrapper`, globalHashId.value, hashId.value]}
              style={style.value}
              {...attrs}
            >
              {child}
            </TransitionGroup>
          </CdkPortal>
        </>
      )
    }
  },
})

const useMessage = (maxCount: ComputedRef<number>) => {
  const messages = ref<MessageOptions[]>([])

  const getCurrIndex = (key: VKey) => {
    return messages.value.findIndex(message => message.key === key)
  }

  const add = (item: MessageOptions) => {
    const currIndex = item.key ? getCurrIndex(item.key) : -1
    if (currIndex !== -1) {
      messages.value.splice(currIndex, 1, item)
      return item.key!
    }

    if (messages.value.length >= maxCount.value) {
      messages.value = messages.value.slice(-maxCount.value + 1)
    }

    const key = item.key ?? uniqueId('ix-message')
    messages.value.push({ ...item, key })
    return key
  }

  const update = (key: VKey, item: MessageOptions) => {
    const currIndex = getCurrIndex(key)
    if (currIndex !== -1) {
      const newItem = { ...messages.value[currIndex], ...item }
      messages.value.splice(currIndex, 1, newItem)
    }
  }

  const destroy = (key: VKey | VKey[]) => {
    const keys = convertArray(key)
    keys.forEach(key => {
      const currIndex = getCurrIndex(key)
      if (currIndex !== -1) {
        const item = messages.value.splice(currIndex, 1)
        callEmit(item[0].onDestroy, key)
      }
    })
  }

  const destroyAll = () => {
    messages.value = []
  }

  const loadContainer = ref(false)

  const open = (options: MessageOptions): MessageRef => {
    const key = add(options)
    const ref = {
      key,
      update: (options: MessageOptions) => update(key, options),
      destroy: () => destroy(key),
    }
    if (!loadContainer.value) {
      loadContainer.value = true
    }
    return ref
  }

  const messageTypes = ['info', 'success', 'warning', 'error', 'loading'] as const
  const [info, success, warning, error, loading] = messageTypes.map(type => {
    return (content: string | VNode | (() => VNodeChild), options?: Omit<MessageOptions, 'type' | 'content'>) =>
      open({ ...options, content, type })
  })

  return { messages, loadContainer, open, info, success, warning, error, loading, update, destroy, destroyAll }
}
