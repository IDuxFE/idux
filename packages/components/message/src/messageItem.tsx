import type { MessageType } from './types'

import { computed, defineComponent, isVNode, onMounted, onUnmounted, onUpdated } from 'vue'
import { PropTypes } from '@idux/cdk/utils'
import { useGlobalConfig } from '@idux/components/config'
import { IxIcon } from '@idux/components/icon'

export default defineComponent({
  name: 'IxMessageItem',
  props: {
    content: PropTypes.oneOfType([PropTypes.string, PropTypes.vNode]).def(''),
    duration: PropTypes.number,
    icon: PropTypes.oneOfType([PropTypes.string, PropTypes.vNode]),
    id: PropTypes.string.isRequired,
    destroyOnHover: PropTypes.bool,
    type: PropTypes.oneOf(['info', 'success', 'warning', 'error', 'loading'] as const).isRequired,
  },
  emits: ['destroy'],
  setup(props, { emit }) {
    const classes = computed(() => {
      const clsPrefix = 'ix-message-item'
      return [clsPrefix, `${clsPrefix}-${props.type}`]
    })

    const config = useGlobalConfig('message')

    const duration = props.duration ?? config.duration
    const destroyOnHover = props.destroyOnHover ?? config.destroyOnHover
    const autoClose = duration > 0

    let timer: number | null = null
    const startTimer = () => {
      timer = setTimeout(() => destroy(), duration)
    }
    const clearTimer = () => {
      if (timer) {
        clearTimeout(timer)
        timer = null
      }
    }
    const destroy = () => {
      clearTimer()
      emit('destroy', props.id)
    }

    const onMouseEnter = () => {
      if (autoClose && !destroyOnHover) {
        clearTimer()
      }
    }
    const onMouseLeave = () => {
      if (autoClose && !destroyOnHover) {
        startTimer()
      }
    }

    onMounted(() => {
      if (autoClose) {
        startTimer()
      }
    })

    onUpdated(() => {
      clearTimer()
      if (autoClose) {
        startTimer()
      }
    })

    onUnmounted(() => clearTimer())

    return { classes, onMouseEnter, onMouseLeave }
  },

  render() {
    const { classes, onMouseEnter, onMouseLeave, content, icon, type } = this
    const iconNode = isVNode(icon) ? icon : useIcon(icon as string, type!)
    const contentNode = isVNode(content) ? content : content
    return (
      <div class={classes} onMouseenter={onMouseEnter} onMouseleave={onMouseLeave}>
        <div class="ix-message-item-content">
          <div class="ix-message-item-custom-content">
            {iconNode}
            <span class="ix-message-item-custom-text">{contentNode}</span>
          </div>
        </div>
      </div>
    )
  },
})

const iconMap = {
  success: 'check-circle',
  error: 'close-circle',
  info: 'info-circle',
  warning: 'exclamation-circle',
  loading: 'loading',
} as const

const useIcon = (icon: string, type: MessageType) => {
  const name = icon ?? iconMap[type]
  return <IxIcon name={name} />
}
