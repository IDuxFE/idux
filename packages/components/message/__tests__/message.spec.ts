import IxMessage from '../src/Message'
import { h } from 'vue'
import { wait } from '@tests'
import { MessageType } from '../src/types'
import { IxIcon } from '@idux/components/icon'

describe('Message', () => {
  let container: HTMLElement
  const getMessageItems = () => container.querySelectorAll('.ix-message-item')
  const messageTypes: MessageType[] = ['info', 'success', 'warning', 'error', 'loading']

  beforeAll(() => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    IxMessage.install({} as any)
    container = document.querySelector('.ix-message-container')!
  })

  beforeEach(() => {
    IxMessage.destroy()
  })

  messageTypes.forEach(type => {
    test(`${type} work`, async () => {
      const content = `This is a ${type} message`
      const open = IxMessage[type]
      open(content)
      await wait(100)

      expect(container.textContent).toContain(content)
      expect(container.querySelector(`.ix-message-item-${type}`)).toBeTruthy()
    })
  })

  test('open work', async () => {
    const content = 'Open message'
    IxMessage.open({ content, type: 'info' })
    await wait(100)

    expect(container.textContent).toContain(content)
  })

  test(`VNode content work`, async () => {
    const content = h('div', { class: 'custom-content' }, 'custom content')
    IxMessage.info(content)
    await wait(200)

    expect(container.textContent).toContain('custom content')
    expect(document.querySelectorAll('.custom-content').length).toBe(1)
  })

  test(`icon work`, async () => {
    const iconName = 'star'
    const iconVNode = h(IxIcon, { name: 'star' })
    IxMessage.info('icon name', { icon: iconName })
    IxMessage.info('icon VNode', { icon: iconVNode })

    await wait(100)

    expect(document.querySelectorAll('.ix-icon-star').length).toBe(2)
  })

  test('destroy work', async () => {
    const instance = IxMessage.info('message')
    await wait(200)

    expect(getMessageItems().length).toBe(1)

    IxMessage.destroy(`not exist id`)
    await wait(100)

    expect(getMessageItems().length).toBe(1)

    IxMessage.destroy(instance.id)
    await wait(100)

    expect(getMessageItems().length).toBe(0)

    IxMessage.info('message1')
    IxMessage.info('message2')
    await wait(100)

    expect(getMessageItems().length).toBe(2)

    IxMessage.destroy()
    await wait(100)

    expect(getMessageItems().length).toBe(0)
  })

  test('instance destroy word', async () => {
    const instance = IxMessage.info('message')
    await wait(100)

    expect(getMessageItems().length).toBe(1)

    instance.destroy()
    await wait(100)

    expect(getMessageItems().length).toBe(0)
  })

  test(`duration work`, async () => {
    IxMessage.success('message', { duration: 1000 })
    await wait(100)

    expect(getMessageItems().length).toBe(1)

    await wait(1000)

    expect(getMessageItems().length).toBe(0)

    IxMessage.success('message', { duration: 0 })

    await wait(100)

    expect(getMessageItems().length).toBe(1)
  })

  test('destroyOnHover work', async () => {
    IxMessage.info('message', { destroyOnHover: true, duration: 200 })
    await wait(100)

    container.querySelector('.ix-message-item')!.dispatchEvent(new MouseEvent('mouseenter'))
    await wait(200)

    expect(getMessageItems().length).toBe(0)

    IxMessage.info('message2', { duration: 200 })
    await wait(100)
    container.querySelector('.ix-message-item')!.dispatchEvent(new MouseEvent('mouseenter'))
    await wait(200)

    expect(getMessageItems().length).toBe(1)

    container.querySelector('.ix-message-item')!.dispatchEvent(new MouseEvent('mouseleave'))
    await wait(250)

    expect(getMessageItems().length).toBe(0)
  })

  test('id work', async () => {
    const id1 = 'id1'
    const id2 = 'id2'
    IxMessage.info(`message1`, { id: id1 })
    IxMessage.info(`message2`, { id: id1 })
    IxMessage.info(`message3`, { id: id2 })

    await wait(100)

    expect(getMessageItems().length).toBe(2)
    expect(container.textContent).not.toContain('message1')
    expect(container.textContent).toContain('message2message3')

    IxMessage.info(`message4`, { id: id2 })
    await wait(100)

    expect(container.textContent).toContain('message4')
  })

  test('onDestroy work', async () => {
    const onDestroy = jest.fn()

    IxMessage.info('message', { onDestroy })

    await wait(3100)

    expect(onDestroy).toBeCalledTimes(1)
  })

  test('maxCount work', async () => {
    new Array(10).fill('message').forEach((content, index) => {
      IxMessage.info(`${content}${index}`)
    })
    await wait(100)

    expect(getMessageItems().length).toBe(5)
  })
})
