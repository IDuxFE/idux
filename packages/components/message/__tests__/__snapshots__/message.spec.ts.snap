// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message render work 1`] = `
"<div class=\\"ix-message ix-message-info\\">
  <div class=\\"ix-message-content\\"><span class=\\"ix-message-content-icon\\"><i class=\\"ix-icon ix-icon-info-circle-filled\\" role=\\"img\\" aria-label=\\"info-circle-filled\\"></i></span><span class=\\"ix-message-content-text\\">This is a message</span></div>
</div>"
`;
